### ISSUE TITLE
Bug: [no-unnecessary-condition] neverNullish inconsistently reported between computed and dot access with `??`

### ISSUE THREAD

**Issue by Grit03** [typescript-eslint/typescript-eslint#12580]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.
- [x] If I used AI assistance to prepare this issue, I have reviewed and followed the [AI Contribution Policy](https://typescript-eslint.io/contributing/ai-policy/).

### Playground Link

https://typescript-eslint.io/play/#ts=6.0.3&fileType=.tsx&code=CYUwxgNghgTiAEYD2A7AzgF3gMyUgXPAErhIzAA8mMAligOYA08KArgLYBGIMAfANwAoQbiQBtAEScoALwkBdeAH4l8AIxDRAOmkzlqjUA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
declare const foo: Record<string, number>;

foo["baz"] ?? 1;
foo.baz ?? 1; // Unnecessary conditional, expected left-hand side of `??` operator to be possibly null or undefined.
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

The two expressions are semantically identical.
Both resolve through the string index signature and both may evaluate to `undefined` at runtime when the key is absent — so they should produce the same result, whichever that is.

### Actual Result

Bracket access is not reported, while dot access is reported as `neverNullish`.
(`noPropertyAccessFromIndexSignature` doesn't affect the types here; when it is off, both syntaxes are legal and typed identically as `number`.)

### Additional Info

From a type perspective, since the `noUncheckedIndexedAccess` option is not enabled, `foo["baz"]` is always inferred as `number`.
So it seems correct that `neverNullish` should be reported.

However, it appears that following the discussion in https://github.com/typescript-eslint/typescript-eslint/issues/6632#issuecomment-1702459452, an exception for the `??` and `??=` operators was introduced in #8249

The exception here means that when the left-hand side of a `??` or `??=` operation is an expression like `foo["baz"]`, it is treated as potentially `undefined`. This is handled by the following condition in `checkNodeForNullish` function:

```typescript
else if (
  !isPossiblyNullish(type) &&
  !(
    node.type === AST_NODE_TYPES.MemberExpression &&
    isNullableMemberExpression(node)
  )
)
``` 
Currently, this exception only applies to computed access.

If this exception is intended, it seems reasonable that dot access like `foo.baz`, which has the same meaning, should also be treated as potentially `undefined`. In that case, `neverNullish` should not be reported for `foo.baz` either.

I'd love to hear any thoughts on this. Once a direction is decided, if this is considered a bug, I'd be happy to open a PR.
